File: daisy/block_bookkeeper.py
import time


class BlockLog:

    def __init__(self, block, stream):
        self.block = block
        self.stream = stream
        self.time_sent = time.time()


class BlockBookkeeper:

    def __init__(self, processing_timeout=None):
        self.processing_timeout = processing_timeout
        self.sent_blocks = {}

    def notify_block_sent(self, block, stream):
        """Notify the bookkeeper that a block has been sent to a client (i.e.,
        a stream to the client)."""

        assert block.block_id not in self.sent_blocks, (
            f"Attempted to send block {block}, although it is already being "
            f"processed by {self.sent_blocks[block.block_id].stream}"
        )

        self.sent_blocks[block.block_id] = BlockLog(block, stream)

    def notify_block_returned(self, block, stream):
        """Notify the bookkeeper that a block was returned."""

        assert block.block_id in self.sent_blocks, (
            f"Block {block} was returned by {stream}, but is not in list "
            "of sent blocks"
        )

        log = self.sent_blocks[block.block_id]
        block.started_processing = log.time_sent
        block.stopped_processing = time.time()

        assert stream == log.stream, (
            f"Block {block} was returned by {stream}, but was sent to" f"{log.stream}"
        )

        del self.sent_blocks[block.block_id]

    def is_valid_return(self, block, stream):
        """Check whether the block from the given client (i.e., stream) is
        expected to be returned from this client. This is to avoid double
        returning blocks that have already been returned as lost blocks, but
        still come back from the client due to race conditions."""

        # block was never sent or already returned
        if block.block_id not in self.sent_blocks:
            return False

        # block is returned by different client than expected
        if self.sent_blocks[block.block_id].stream != stream:
            return False

        return True

    def get_lost_blocks(self):
        """Return a list of blocks that were sent and are lost, either because
        the stream to the client closed or the processing timed out. Those
        blocks are removed from the sent-list with the call of this
        function."""

        lost_block_ids = []
        for block_id, log in self.sent_blocks.items():

            # is the stream to the client still alive?
            if log.stream.closed():
                lost_block_ids.append(block_id)

            # did processing time out?
            if self.processing_timeout is not None:
                if time.time() - log.time_sent > self.processing_timeout:
                    lost_block_ids.append(block_id)

        lost_blocks = []
        for block_id in lost_block_ids:

            lost_block = self.sent_blocks[block_id].block
            lost_blocks.append(lost_block)
            del self.sent_blocks[block_id]

        return lost_blocks

File: gerbera-core/src/block_bookkeeper.rs
use crate::block::{Block, BlockId};
use std::collections::HashMap;
use std::net::SocketAddr;
use std::time::Instant;
use tracing::debug;

/// Tracks in-flight blocks to detect lost blocks (stream closed or timeout).
struct BlockLog {
    block: Block,
    client_addr: SocketAddr,
    time_sent: Instant,
}

pub struct BlockBookkeeper {
    processing_timeout: Option<std::time::Duration>,
    sent_blocks: HashMap<BlockId, BlockLog>,
    /// Track which client addresses have disconnected so we can detect lost blocks.
    closed_clients: std::collections::HashSet<SocketAddr>,
}

impl BlockBookkeeper {
    pub fn new(processing_timeout: Option<std::time::Duration>) -> Self {
        Self {
            processing_timeout,
            sent_blocks: HashMap::new(),
            closed_clients: std::collections::HashSet::new(),
        }
    }

    /// Record that a block was sent to a client.
    pub fn notify_block_sent(&mut self, block: Block, client_addr: SocketAddr) {
        debug!(block_id = %block.block_id, %client_addr, "block sent to client");
        self.sent_blocks.insert(
            block.block_id.clone(),
            BlockLog {
                block,
                client_addr,
                time_sent: Instant::now(),
            },
        );
    }

    /// Record that a block was returned by a client. Returns the time
    /// the block spent in the worker (`Instant::now() - time_sent`)
    /// when the return is valid, otherwise `None`.
    pub fn notify_block_returned(
        &mut self,
        block: &Block,
        client_addr: SocketAddr,
    ) -> Option<std::time::Duration> {
        if let Some(log) = self.sent_blocks.get(&block.block_id) {
            if log.client_addr == client_addr {
                let elapsed = log.time_sent.elapsed();
                self.sent_blocks.remove(&block.block_id);
                return Some(elapsed);
            }
            debug!(
                block_id = %block.block_id,
                expected = %log.client_addr,
                actual = %client_addr,
                "block returned by wrong client"
            );
        }
        None
    }

    /// Check whether this return is valid (block was sent to this client and
    /// hasn't already been returned).
    pub fn is_valid_return(&self, block: &Block, client_addr: SocketAddr) -> bool {
        if let Some(log) = self.sent_blocks.get(&block.block_id) {
            log.client_addr == client_addr
        } else {
            false
        }
    }

    /// Mark a client address as disconnected.
    pub fn notify_client_disconnected(&mut self, client_addr: SocketAddr) {
        self.closed_clients.insert(client_addr);
    }

    /// Return blocks that are lost: either the client disconnected or
    /// processing timed out. Removes them from the sent list.
    pub fn get_lost_blocks(&mut self) -> Vec<Block> {
        let now = Instant::now();
        let mut lost_ids = Vec::new();

        for (block_id, log) in &self.sent_blocks {
            if self.closed_clients.contains(&log.client_addr) {
                lost_ids.push(block_id.clone());
                continue;
            }
            if let Some(timeout) = self.processing_timeout {
                if now.duration_since(log.time_sent) > timeout {
                    lost_ids.push(block_id.clone());
                }
            }
        }

        let mut lost_blocks = Vec::new();
        for id in lost_ids {
            if let Some(log) = self.sent_blocks.remove(&id) {
                lost_blocks.push(log.block);
            }
        }
        lost_blocks
    }
}
